test-suite/testkit/src/api.rs

```diff
@@ -378,3 +386,75 @@ fn create_test_server(aggregator: ApiAggregator) -> TestServer {
     info!("Test server created on {}", server.addr());
     server
 }
+
+/// The custom implementation of the test server, because there is an error in the default
+/// implementation. It does not wait for the http server thread to complete during drop.
+struct TestServer {
+    addr: net::SocketAddr,
+    backend: Addr<Server>,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl TestServer {
+    /// Start new test server with application factory
+    fn with_factory<F, H>(factory: F) -> Self
+    where
+        F: Fn() -> H + Send + Clone + 'static,
+        H: IntoHttpHandler + 'static,
+    {
+        let (tx, rx) = mpsc::channel();
+
+        // run server in separate thread
+        let handle = thread::spawn(move || {
+            let sys = System::new("actix-test-server");
+            let tcp = net::TcpListener::bind("127.0.0.1:0").unwrap();
+            let local_addr = tcp.local_addr().unwrap();
+
+            let srv = HttpServer::new(factory)
+                .disable_signals()
+                .listen(tcp)
+                .keep_alive(5)
+                .workers(1)
+                .start();
+
+            tx.send((System::current(), local_addr, srv)).unwrap();
+            sys.run();
+        });
+
+        let (system, addr, backend) = rx.recv().unwrap();
+        System::set_current(system);
```

Comment:
Couldn't this call be avoided if `System` is retained within `TestServer`? IMO, setting a `System` for the test thread may have unintended side effects.

Reply:
I just copypasted this code from the `actix_web::TestServer`.